astropy/coordinates/tests/test_sky_coord.py

```diff
@@ -668,6 +673,41 @@ def test_to_string():
         assert with_kwargs == wrap('+01h02m03.000s +01d02m03.000s')
 
 
+def test_to_table():
```

Comment:
There should also be round-trip tests that ``SkyCoord.guess_from_table(sc.to_table())`` works as expected.

Reply:
Round-tripping would certainly be nice, but I'm not sure there should be any expectations. I have the impression that `guess_from_table()` is named the way it is convey that it might not always work.

Reply:
For tables generated "in the wild" it is certainly the case that ``SkyCoord.guess_from_table`` might not work, but for (new, not updated) Tables generated by ``SkyCoord.to_table`` I think the expectation is that it MUST work. 
The ``SkyCoord.to_table`` has a very clear prescription for making and naming columns. If round-tripping doesn't work, then  ``SkyCoord.guess_from_table`` needs to be fixed. Implementing such improvements are out of scope of this PR, but I think a roundtrip test is not.

Reply:
It seems to me that round-tripping can be expected to always work only if `guess_from_table()` can always recognize if its input is a converted `SkyCoord` or not, and that seems quite complicated. It is also possible to be less ambitious and only require round-tripping to work for some subset of `SkyCoord` objects, but specifying such a subset would be outside the scope of this pull request.

Reply:
It's certainly true that some of the features of ``to_table`` are not currently covered by ``guess_from_table`` — eg  obstime. And achieving parity is out of scope here.
If @mhvk / @adrn / @taldcroft agree that no test of ``SkyCoord.guess_from_table(sc.to_table()) == sc`` should be in this PR than an [issue](https://github.com/astropy/astropy/issues/new?assignees=&labels=&template=feature_request.md) should be made to *start* these tests. Regardless, I do think a [new issue](https://github.com/astropy/astropy/issues/new?assignees=&labels=&template=feature_request.md) is merited to highlight the myriad ways in which ``SkyCoord.guess_from_table`` needs to be improved so that ``SkyCoord.guess_from_table(sc.to_table()) == sc`` works. I can make this issue, or you can. It's your PR and good work here, so the option's yours.

Reply:
I think I overestimated how complicated making round-tripping work would be. If a table has been produced by `SkyCoord.to_table()` then its metadata will contain `frame` and `representation_type`. It is possible to look up what the parameters of a `SkyCoord` with this `frame` and `representation_type` combination are called and looking up which of those are actually present in the table is very simple.  Any possible coordinate frame parameters can be listed with `frame_transform_graph.frame_attributes.keys()`, so any metadata keys not listed there are not relevant for reconstructing the coordinate frame. No guesswork required. So I agree that it is reasonable to expect that _any_ `SkyCoord` could be round-tripped and I no longer think that implementing it would be overly complicated.

Having established that it should be possible to round-trip any `SkyCoord` it becomes apparent that testing such round-tripping would be analogous to testing round-tripping a `SkyCoord` as a mixin column in `Table` serialization (because that is supposed to work for any `SkyCoord` too). So the already existing tests just need to be adapted for the round-tripping here and there shouldn't be any need to write tests from scratch. But I still believe adapting the aforementioned tests would naturally be a part of the PR that would update `guess_from_table()` and not this PR.

Reply:
> I think I overestimated how complicated making round-tripping work would be.

The best type of estimation. 👍 

> No guesswork required. So I agree that it is reasonable to expect that any SkyCoord could be round-tripped...

Great. A good example of [Postel's Law](https://en.wikipedia.org/wiki/Robustness_principle).

> testing such round-tripping would be analogous to testing round-tripping a SkyCoord as a mixin column in Table serialization (because that is supposed to work for any SkyCoord too). So the already existing tests just need to be adapted for the round-tripping here and there shouldn't be any need to write tests from scratch.

Even better. 

> But I still believe adapting the aforementioned tests would naturally be a part of the PR that would update ``guess_from_table()`` and not this PR.

The above shows you've thought deeply on how this can be implemented and if there's a pre-v5.0 timeline for ensuring round-tripping, then I have no objections. My initial hesitation / insistence was predicated on the uncertainty in the amount of effort required to get this all to work. Making a test here would highlight the myriad ways in which ``guess_from_table()`` doesn't yet support round-tripping. But if this is already quantified then I have no objections.